spec.md

```diff
@@ -0,0 +1,269 @@
+Here we define the fields we need from it.
+
+```
+    WasmVM ::= { stack : InstructionStack }
+```
+
+An `InstructionStack` consists of a space-separate list of `Instruction`.
+
+```
+    InstructionStack ::= ".Stack"
+                       | InstructionStack Instruction
+```
+
+### The EEI
+
+The EEI contains fields relevant to VMs used for executing contracts on the Ethereum blockchain.
+
+**TODO**: In the definition of `EEI.useGas`, the field `gas` is treated like an `i64.const`, but here it's an `Int256`.
+          How do we reconcile these?
+          In EVM, it's an `Int256` in general.
+
+```
+    EEI ::= { statusCode : StatusCode
+            , accounts   : Accounts
+            , gas        : Int256
+            }
+```
+
+The `StatusCode` is defined by the [EVM-C API Standard](https://github.com/ethereum/evmc).
+See the repository for an explanation of each status code.
+
+```
+    StatusCode ::= "EVM_SUCCESS"
+                 | "EVM_FAILURE"
+                 | "EVM_INVALID_INSTRUCTION"
+                 | "EVM_UNDEFINED_INSTRUCTION"
+                 | "EVM_OUT_OF_GAS"
+                 | "EVM_BAD_JUMP_DESTINATION"
+                 | "EVM_STACK_OVERFLOW"
+                 | "EVM_STACK_UNDERFLOW"
+                 | "EVM_INVALID_MEMORY_ACCESS"
+                 | "EVM_REVERT"
+                 | "EVM_STATIC_MODE_ERROR"
+                 | "EVM_REJECTED"
+                 | "EVM_INTERNAL_ERROR"
+```
+
+Accounts store the four relevant fields for accounts on the Ethereum network.
+
+```
+    AccountId ::= Int160
+    Accounts  ::= Map{AccountId, Account}
+    Account   ::= { balance : Int256
+                  , code    : ByteString
+                  , storage : Map{Int256, Int256}
+                  , nonce   : Int256
+                  }
+```
+
+Initialisation of a contract
+----------------------------
+
+**TODO**
+
+Execution of the contract entry point
+-------------------------------------
+
+**TODO**
+
+Host functions available to the contract
+----------------------------------------
+
+The EEI specifies several host functions available to the user.
+Conceptually, it is defined as a wasm module which is available while executing ewasm contracts.
+The definition of the functions is *opaque*, meaning they are provided in terms of their update to the state and not in terms of wasm code.
+In the future, they may become more transparent as suitable wasm code is found to represent them.
+
+### EEI Module
+
+The function signatures are provided here in wast format:
+
+```wast
+    (module
+        (func (export "EEI.useGas")         (param i64)                 (result))
+        (func (export "EEI.call")           (param i64 i32 i32 i32 i32) (result i32))
+        (func (export "EEI.returnDataCopy") (param i32 i32 i32)         (result))
+    )
+```
+
+### `(call EEI.useGas (i64.const g))`
+
+Function to deduct the correct amount of gas for execution of an ewasm contract.
+
+1.  Load the value `i64.const G` from `EEI.gas`.
```

Comment:
The use of `g` and `G` here is a little confusing. It took me a moment to figure this out. I prefer clearly-named variables, in this case something like, `gasToUse` and `GAS_REMAINING`.

It's also a little unclear at first glance that `EEI.gas` is some sort of global (is this even the right terminology?), since `EEI.useGas` is a function. Use of a sigil like `$EEI.gas` could make this clearer.

Reply:
+1 for using clearly-named vars

Reply:
Sounds good, will do.